Switch to use libchart approach to avoid code duplication

ekarlso opened the pull request.

ah9828 commented on dotnet/templates/NOTES.txt:
```diff
@@ -4,16 +4,16 @@
 {{- end }}
 {{- else if contains "NodePort" .Values.service.type }}
-  export NODE_PORT=$(kubectl get --namespace {{ .Release.Namespace }} -o jsonpath="{.spec.ports[0].nodePort}" services {{ template "dotnet.name" . }})
+  export NODE_PORT=$(kubectl get --namespace {{ .Release.Namespace }} -o jsonpath="{.spec.ports[0].nodePort}" services {{ template "java.name" . }})
```
```suggestion
  export NODE_PORT=$(kubectl get --namespace {{ .Release.Namespace }} -o jsonpath="{.spec.ports[0].nodePort}" services {{ template "libchart.name" . }})
```

ah9828 commented on golang/templates/NOTES.txt:
```diff
@@ -4,16 +4,16 @@
 {{- end }}
 {{- else if contains "NodePort" .Values.service.type }}
-  export NODE_PORT=$(kubectl get --namespace {{ .Release.Namespace }} -o jsonpath="{.spec.ports[0].nodePort}" services {{ template "golang.fullname" . }})
+  export NODE_PORT=$(kubectl get --namespace {{ .Release.Namespace }} -o jsonpath="{.spec.ports[0].nodePort}" services {{ template "java.name" . }})
```
```suggestion
  export NODE_PORT=$(kubectl get --namespace {{ .Release.Namespace }} -o jsonpath="{.spec.ports[0].nodePort}" services {{ template "libchart.name" . }})
```

ah9828 commented on nodejs/templates/NOTES.txt:
```diff
@@ -4,16 +4,16 @@
 {{- end }}
 {{- else if contains "NodePort" .Values.service.type }}
-  export NODE_PORT=$(kubectl get --namespace {{ .Release.Namespace }} -o jsonpath="{.spec.ports[0].nodePort}" services {{ template "nodejs.name" . }})
+  export NODE_PORT=$(kubectl get --namespace {{ .Release.Namespace }} -o jsonpath="{.spec.ports[0].nodePort}" services {{ template "java.name" . }})
```
```suggestion
  export NODE_PORT=$(kubectl get --namespace {{ .Release.Namespace }} -o jsonpath="{.spec.ports[0].nodePort}" services {{ template "libchart.name" . }})
```

ah9828 commented on web/templates/NOTES.txt:
```diff
@@ -4,16 +4,16 @@
 {{- end }}
 {{- else if contains "NodePort" .Values.service.type }}
-  export NODE_PORT=$(kubectl get --namespace {{ .Release.Namespace }} -o jsonpath="{.spec.ports[0].nodePort}" services {{ template "web.fullname" . }})
+  export NODE_PORT=$(kubectl get --namespace {{ .Release.Namespace }} -o jsonpath="{.spec.ports[0].nodePort}" services {{ template "java.name" . }})
```
```suggestion
  export NODE_PORT=$(kubectl get --namespace {{ .Release.Namespace }} -o jsonpath="{.spec.ports[0].nodePort}" services {{ template "libchart.name" . }})
```

ah9828 reviewed (changes requested):
sjekk også i linje 7 på java/templates/NOTES.txt

Haavare commented:
When the comments are resolved this looks good

ah9828 reviewed (changes requested):
@ekarlso Etter å ha sjekket nøyere etter så lurer jeg på om du generer duplikater siden du både har templates mappen for hvert språk og har tillegg templates mappen med samme tingene i libchart

Haavare commented:
I can't se any duplicated `*.yaml` files, the `_*.tpl` files will not create files of their own.

Haavare commented:
@ekarlso you have to run `helm dependency update` to import the libchart tar file into `/charts`

Haavare commented:
here is a example of `helm template libchart-test ./dotnet/` after importing the libchar  tar file:
```
---
# Source: dotnet/templates/service-account.yaml
apiVersion: v1
kind: ServiceAccount
metadata:
  name: libchart-test-dotnet
  namespace: marine-resources-prod
  labels:
    helm.sh/chart: dotnet-9.0.0

    app.kubernetes.io/name: libchart-test
    app.kubernetes.io/instance: libchart-test

    app.kubernetes.io/managed-by: Helm
    app.kubernetes.io/part-of: libchart-test
    app.kubernetes.io/component: libchart-test
---
# Source: dotnet/templates/service.yaml
apiVersion: v1
kind: Service
metadata:
  name: libchart-test
  namespace: marine-resources-prod
  labels:
    helm.sh/chart: dotnet-9.0.0

    app.kubernetes.io/name: libchart-test
    app.kubernetes.io/instance: libchart-test

    app.kubernetes.io/managed-by: Helm
    app.kubernetes.io/part-of: libchart-test
    app.kubernetes.io/component: libchart-test
spec:
  type: ClusterIP
  ports:
    - port: 80
      targetPort: 8080
      protocol: TCP
      name: http

  selector:

    app.kubernetes.io/name: libchart-test
    app.kubernetes.io/instance: libchart-test
---
# Source: dotnet/templates/deployment.yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: libchart-test
  namespace: marine-resources-prod
  labels:
    helm.sh/chart: dotnet-9.0.0

    app.kubernetes.io/name: libchart-test
    app.kubernetes.io/instance: libchart-test

    app.kubernetes.io/managed-by: Helm
    app.kubernetes.io/part-of: libchart-test
    app.kubernetes.io/component: libchart-test
spec:
  replicas: 1
  selector:
    matchLabels:

      app.kubernetes.io/name: libchart-test
      app.kubernetes.io/instance: libchart-test
  template:
    metadata:
      labels:
        helm.sh/chart: dotnet-9.0.0

        app.kubernetes.io/name: libchart-test
        app.kubernetes.io/instance: libchart-test

        app.kubernetes.io/managed-by: Helm
        app.kubernetes.io/part-of: libchart-test
        app.kubernetes.io/component: libchart-test
    spec:

      serviceAccountName: libchart-test-dotnet
      securityContext:
        runAsNonRoot: true
      containers:
        - name: dotnet
          image: ":"
          imagePullPolicy: Always
          ports:
            - name: http
              containerPort: 8080
              protocol: TCP

          livenessProbe:
            httpGet:
              path: /
              port: 8080
            initialDelaySeconds: 15
            timeoutSeconds: 15
            periodSeconds: 15
          readinessProbe:
            httpGet:
              path: /
              port: 8080
            initialDelaySeconds: 15
            timeoutSeconds: 15
            periodSeconds: 15
          env:
          resources:

            {}
      affinity:
        podAntiAffinity:
          preferredDuringSchedulingIgnoredDuringExecution:
            - weight: 1
              podAffinityTerm:
                topologyKey: "kubernetes.io/hostname"
                labelSelector:
                  matchLabels:

                    app.kubernetes.io/name: libchart-test
                    app.kubernetes.io/instance: libchart-test
---
# Source: dotnet/templates/servicemonitor.yaml
apiVersion: monitoring.coreos.com/v1
kind: ServiceMonitor
metadata:
  name: libchart-test
  namespace: marine-resources-prod
  labels:
    helm.sh/chart: dotnet-9.0.0

    app.kubernetes.io/name: libchart-test
    app.kubernetes.io/instance: libchart-test

    app.kubernetes.io/managed-by: Helm
    app.kubernetes.io/part-of: libchart-test
    app.kubernetes.io/component: libchart-test
    prometheus: default
spec:
  selector:
    matchLabels:

      app.kubernetes.io/name: libchart-test
      app.kubernetes.io/instance: libchart-test
  endpoints:
  - path: /metrics
    interval: 15s
    targetPort: 8080
    honorLabels: true
```
I do not know if this is a good example

Haavare commented:
I think this is a good examle of a situation where Architecture Decision Records would make sense to have:
https://cognitect.com/blog/2011/11/15/documenting-architecture-decisions

At the moment I am not sure why libchart is considered less complex than maintaining separate charts

ekarlso commented:
> I think this is a good examle of a situation where Architecture Decision Records would make sense to have:
> https://cognitect.com/blog/2011/11/15/documenting-architecture-decisions
> 
> At the moment I am not sure why libchart is considered less complex than maintaining separate charts

Because now implementing a change or a common component is at least common across them all ;)

Whether we should actually have seperate charts for java, dotnet, or whatever language you think if is a better valid question in my mind.

Haavare commented:
Yes, I actually thought you all decided only the `web` chart was necessary two years ago

Haavare commented:
There are `namespace` in some but not all templates, I do not think we need them, but could we be consistent either or.

ekarlso commented:
> There are `namespace` in some but not all templates, I do not think we need them, but could we be consistent either or.

Removed namespaces and now and added policy to keep it removed :

Haavare commented on web/values.yaml:
```diff
@@ -151,8 +175,8 @@ cert:
   issuerKind: ClusterIssuer
-  # secretName: web-appname-tls
-  # duration: 1440h
+  # secretName: java-appname-tls
```
It is not that important but you have this java comment in all the charts

Haavare reviewed (commented):
You check if it is a statefulset twice, on include and define. I think that is fine.
I am not familiar with the Istio templates, so I just checke spelling.
Otherwise did you see my comment on labelSelectors above? That is the only thing I have left

ekarlso commented on .gitmodules:
```diff
@@ -2,5 +2,5 @@
-	path = istio-bookinfo
-	url = https://github.com/evry-ace/helm-istio-bookinfo.git
+       path = istio-bookinfo
+       url = https://github.com/evry-ace/helm-istio-bookinfo.git
```
Done automatically I believe :thinking:

ekarlso commented on libchart/templates/_helpers.tpl:
```diff
@@ -0,0 +1,69 @@
+{{- define "libchart.labels" -}}
+helm.sh/chart: {{ include "libchart.chart" . }}
+{{ include "libchart.selectorLabels" . }}
+{{ include "libchart.versionLabel" . }}
```
:+1:

ekarlso commented on golang/templates/cm-certificate.yaml:
```diff
@@ -1,36 +1,3 @@
-
-{{- if and .Values.cert.enabled ($hosts) }}
-
-apiVersion: {{ .Values.certmanagerApiVersion }}
```
Yes, breaking change because cert-manager v1 is introduced long long time ago. Also there's a better way to do apiVersion detection atm.

ekarlso commented on dotnet/templates/NOTES.txt:
```diff
@@ -4,16 +4,16 @@
 {{- end }}
 {{- else if contains "NodePort" .Values.service.type }}
-  export NODE_PORT=$(kubectl get --namespace {{ .Release.Namespace }} -o jsonpath="{.spec.ports[0].nodePort}" services {{ template "dotnet.name" . }})
+  export NODE_PORT=$(kubectl get --namespace {{ .Release.Namespace }} -o jsonpath="{.spec.ports[0].nodePort}" services {{ template "java.name" . }})
```
This is already done.

ekarlso commented:
> You check if it is a statefulset twice, on include and define. I think that is fine.
> I am not familiar with the Istio templates, so I just checke spelling.
> Otherwise did you see my comment on labelSelectors? That is the only thing I have left, great work!

Where's this?

ah9828 commented:
Good work @ekarlso!

Haavare commented:
> > You check if it is a statefulset twice, on include and define. I think that is fine.
> > I am not familiar with the Istio templates, so I just checke spelling.
> > Otherwise did you see my comment on labelSelectors? That is the only thing I have left, great work!
> 
> Where's this?

Looks like you have changed it, all good
